src/Entities/Entity.cpp

```diff
@@ -1522,13 +1550,23 @@ bool cEntity::DoMoveToWorld(cWorld * a_World, bool a_ShouldSendRespawn, Vector3d
 	}
 
 	// Queue add to new world
-	a_World->AddEntity(this);
 	cWorld * OldWorld = cRoot::Get()->GetWorld(GetWorld()->GetName());  // Required for the hook HOOK_ENTITY_CHANGED_WORLD
 	SetWorld(a_World);
 
 	// Entity changed the world, call the hook
 	cRoot::Get()->GetPluginManager()->CallHookEntityChangedWorld(*this, *OldWorld);
 
+	OldWorld->QueueTask([this, ParentChunk, a_World](cWorld & a_OldWorld)
+	{
+		LOGD("Warping entity #%i (%s) from world (%s) to (%s). Source chunk: (%d, %d) ",
+			this->GetUniqueID(), this->GetClass(),
+			a_OldWorld.GetName().c_str(), a_World->GetName().c_str(),
+			ParentChunk->GetPosX(), ParentChunk->GetPosZ()
+		);
+		ParentChunk->RemoveEntity(this);
+		ParentChunk->MarkDirty();
+		a_World->AddEntity(this);
+	});
```

Comment:
The hook call for EntityChangedWorld should be moved here, the hook description says "the entity is already in the new world"